0001-job-worker.md

```diff
@@ -0,0 +1,234 @@
+### gRPC
+
+In this PR there is a `worker.proto` file specifying the RPC service and it's messages. These effectively wrap the golang libraries functionality. A gRPC server will be developed importing the library and offering the endpoints to the CLI client.
+
+#### Streaming approach
+
+When starting a job the cmd.Exec will map the STDOUT and STDERR to a file. When the client requests the output stream, the gRPC endpoint can then tail the contents of the file and pipe it through the gRPC response stream to the client. The advantage to this approach is simplicity, the go standard library then handles concurrent writing of STDOUT and STDERR to the file. Concurrent reading then becomes as easy as tailing the file and piping the response through the gRPC stream. Since level 4 did not need resource isolation for mount and PID namespaces, this has not been implemented, but this would need to be done in production to isolate the log of each Job. The disadvantages of this method is that it does not handle log rotations, in the event the process is very long lived or produces many logs. In this case the stream may end and the user could be required to cancel and re run the command.
+
+### Process execution life cycle
+
+Together with this design doc there are two UML sequence diagrams showing a happy and un happy scenario.
+
+![unhappy.plantuml](unhappy.jpg "Unhappy Scenario")
+
+In the unhappy scenario a gRPC client sends a start request to the server. The server then uses the library to first create a cgroup to execute the job, adds the resource control parameters in the appropriate interface files for the cgroup controllers, then starts the job's process. This process is not the client's command, but a simple process that adds it's own PID to the cgroup. This prevents a race condition where the client's command is executed then the PID is found and added, since the command may already abuse resources before the cgroup has a chance to control them. Once the PID is added the command is run in the process. After this the server responds to the client. In this un happy scenario, we assume the command fails quickly and returns an error code. The client then uses the ID in the response to query the status of the job. Since the server has access to the underlying exec.Cmd and therefore os.Process we can return the status along with the return code to the user with the Running bool as false.
+
+![happy.plantuml](happy.jpg "Happy Scenario")
+
+The happy scenario is the same for starting a job. Here we assume the command has not yet returned and is executing, so the status response returns Running true. Therefore the client can now use the ID again to stream the logs of the job. After streaming for a period of time the client can cancel the stream. If the client then decides to stop the job it can send the ID to the server and the server can forcefully close the process using os.Process.Kill. Once killed the cgroup is deleted and a response sent to the client.
```

Comment:
How will this work for long-running processes that need to be stopped gracefully by the client?

Reply:
I've updated the description now to use the SIGTERM signal isntead of forcefully closes the process